### DISCUSSION TITLE
Cannot connect to docker

### DISCUSSION THREAD

**Discussion by nextlogic-ono** [SquirrelCorporation/SquirrelServersManager#456] [Q&A]
I've set up SSM on a machine and added another device on the same network and agent seems to be fine, but I cannot find any containers on that host.

The document says, it should be automatically found but what is the criteria for this to work?
SSM is installed on 10.0.0.1 in a docker and the target machine is 10.0.0.2, which I can SSH into from `squirrelserversmanager-server-1` container and the agent is deployed there but when I try the connection test, the docker part never seems to do anything.

I've tried both `paramiko` and `ssh` as connection method and the connecting user can use the docker command as it belongs to `docker` group as well as being able to `sudo` to root.

On an added note, I also cannot do `Connect` to host via its SSH web interface, despite ansible being able to run with the given credential.

Please let me know what I can try.

**Comment by SquirrelDeveloper** [comment]
Hi @windware-ono . Thanks for the report
You can run a "Connection test" in Inventory / <your device> / Connection Test -> "Run connection tests".
Additionally, please provide the server logs (In the left panel "Logs"). You can filter on either module= "SsmSshDockerAgent" or "Watcher"

**Comment by nextlogic-ono** [comment]
I've tried the tests.
Thanks for checking.

Connection test (Docker connection test has the indicator moving indefinitely)
<img width="521" alt="ssm-connection-tests" src="https://github.com/user-attachments/assets/66f20456-395f-435a-9cad-8f414a7a7e35">

Server logs (SsmSshDockerAgent)
<img width="1133" alt="ssm-server-logs-SsmSshDockerAgent" src="https://github.com/user-attachments/assets/1fadbb0d-07e0-46e4-aa85-6c9d2ad6ab1b">

Server logs (Watcher)
![ssm-server-logs-Watcher](https://github.com/user-attachments/assets/f8fd9835-0df8-42ef-9ae8-d24af3d928e1)

**Comment by SquirrelDeveloper** [comment]
Unfortunately, there is no info here that can help.
I am gonna publish v0.1.22 with a timeout on the check docker connection, so we may get more infos

**Comment by nextlogic-ono** [comment]
I re-ran the connection test and looked at the logs without filtering with a module and this is what I got. Perhaps, this could mean something?

![ssm-connection-test-error](https://github.com/user-attachments/assets/6669ee00-b02a-445f-a38e-4b867b4dd1a2)

**Reply by SquirrelDeveloper** [reply]
Unfortunately not, but the new version is released I encourage you to test it. (I recommend starting from a clean install)

**Comment by nextlogic-ono** [comment]
New version with a clean install didn't do much different. `Watcher` log was slightly different but the errors were same with "request entity too large" and 500 errors.

<img width="1120" alt="Screenshot 2024-11-15 at 2 15 34" src="https://github.com/user-attachments/assets/b9eb3266-8d85-4477-a767-b31ea0f191e6">

**Reply by SquirrelDeveloper** [reply]
Alright. I am gonna implement a small diagnostic module so we can have more infos

**Reply by SquirrelDeveloper** [reply]
What kind of connection to the device do you use? Private keys, user/pwd, ... ?
Is your docker.sock protected with keys?

**Reply by nextlogic-ono** [reply]
I'm connecting to the target machine with user/password who has access to `docker.sock` on the target machine. `docker.sock` is not protected other than being `660` permission by its default.

**Reply by nextlogic-ono** [reply]
When I get back, I'll try with stock Ubuntu 24.04 and see how it will do, though I don't believe I'm doing anything special with my set up.

**Comment by nextlogic-ono** [comment]
I've spun up 2 new Ubuntu 24.04 and only installed Docker on both machines with a few test containers and installed SSM on the first machine.

It now did show up containers but now it said that it couldn't find playbooks and couldn't do the pinging like on the screenshot.

Installing the agent also failed saying the playbook cannot be found.
I cleaned SSM data folder and restarted SSM but it was still having the same error.

<img width="781" alt="ssm-playbook-not-found1" src="https://github.com/user-attachments/assets/b292d97f-e2b8-48d6-b4cf-64c1826c8ad0">
<img width="1059" alt="ssm-playbook-not-found2" src="https://github.com/user-attachments/assets/173f3521-5cc9-4dee-a29b-91be0c679fbe">

**Comment by nextlogic-ono** [comment]
I no longer saw the errors posted before in the logs, except for the logs that say playbooks aren't found.

**Comment by SquirrelDeveloper** [comment]
It seems there is a bug here. Could you try to reset the playbooks module here? 
<img width="1618" alt="Screenshot 2024-11-17 at 08 55 41" src="https://github.com/user-attachments/assets/9b79501a-dd83-4fbb-a0a0-bee8390fa38b">

**Reply by SquirrelDeveloper** [reply]
New version is being released has we speak, hopefully fixing it!

**Comment by nextlogic-ono** [comment]
On a stock Ubuntu 24.04, it worked fine for both the pinging and the docker connection with a clean install of SSM.
But with my own servers, removing the data and doing a clean install of SSM didn't fix the docker connection with the latest release.
Let me know if there's anything I should check.

Also, I found several problems but let me know if I should get these into separate issues.

- I cannot see 'Live Logs' of the containers, despite having it listed as `running` and I checked that they're running on the host.
- Clicking on `Stop` will not immediately give feedback and notification only shows up after some seconds when the container actually stopped, making it confusing whether if any action is being taken on click.
- When the container is actually stopped, the list still says `Running`.
- The port of the link says `undefined`, invalidating the link. (Probably not good to link to it directly in case it is hosted under reverse proxy.)
<img width="453" alt="Screenshot 2024-11-17 at 18 22 51" src="https://github.com/user-attachments/assets/17d7da80-17b4-4b97-8ebb-5cb9a3a2b08c">

**Reply by SquirrelDeveloper** [reply]
Thank for your feedbacks, those are really usefull. I will fix them along the way

**Comment by nextlogic-ono** [comment]
I did another testing by spawning a new Ubuntu 24.04 instance and first tried to connect to the server with `root` and both pinging and docker connection succeeds.

Next, I tried to create a new user and gave it sudo access without password.
It no longer can do the pinging nor docker connection, despite being able to `sudo`.

Next, I put the new user into `docker` group and docker connection succeeded but the pinging still doesn't work.

Are you testing with `root` user connection?

**Reply by SquirrelDeveloper** [reply]
Hi! 
Unless I misundertood, it works on ubuntu with a simple user, even without sudo access.
```shell
# Create a new user
sudo adduser testuser
sudo usermod -aG docker testuser
# Ensure docker access
docker ps
```
Fill in SSM 

<img width="894" alt="Screenshot 2024-11-22 at 16 10 24" src="https://github.com/user-attachments/assets/2a90f63b-0d9d-4c55-864b-70ee99d9a444">

Both docker & Ansible ping are working.

**Reply by nextlogic-ono** [reply]
I was expecting that `sudo` would give the regular user access to `/run/docker.sock`, thus was suffice to allow docker connection as I would expect on terminal without having the user in `docker` group.

I'm not sure why the regular user with `sudo` access can no longer get the pinging to work...

Also, on my own server, having the user in `docker` group nor `root` user still won't make the docker connection successful and I haven't yet figured out how it's different from a stock Ubuntu install.

**Reply by SquirrelDeveloper** [reply]
SSM use sudo through Ansible only when necessary, meaning when a playbook has a become: true option.
I have added a small tool to help debug the different connexion issues in inventory/<device>/Configuration
<img width="906" alt="Screenshot 2024-11-22 at 16 27 42" src="https://github.com/user-attachments/assets/e3578943-d6d1-47b5-a97a-c607523f4b45">
